apps/src/lib/kits/maker/CircuitPlaygroundBoard.js

```diff
@@ -0,0 +1,197 @@
+process.hrtime = require('browser-process-hrtime');
+
+/** @const {number} serial port transfer rate */
+const SERIAL_BAUD = 57600;
+
+const J5_CONSTANTS = {
+  INPUT: 0,
+  OUTPUT: 1,
+  ANALOG: 2,
+  PWM: 3,
+  SERVO: 4
+};
+
+/**
+ * Controller interface for an Adafruit Circuit Playground board using
+ * Circuit Playground Firmata firmware.
+ */
+export default class CircuitPlaygroundBoard extends EventEmitter {
+  constructor(portName) {
+    super();
+
+    /** @private {string} a port identifier, e.g. "/dev/ttyACM0" */
+    this.portName_ = portName;
+
+    /** @private {five.Board} A johnny-five board controller */
+    this.fiveBoard_ = null;
+
+    /** @private {Object} Map of component controllers */
+    this.prewiredComponents_ = null;
+  }
+
+  /**
+   * Open a connection to the board on its configured port.
+   * @return {Promise}
+   */
+  connect() {
+    return Promise.resolve()
+        .then(() => this.connectToFirmware())
+        .then(() => this.initializeComponents())
+        .then(() => this.initializeEventForwarding());
+  }
+
+  /**
+   * Open the serial port connection and create a johnny-five board controller.
+   * Exposed as a separate step here for the sake of the setup page; generally
+   * recommended to just call connect(), above.
+   * @return {Promise}
+   */
+  connectToFirmware() {
+    return new Promise((resolve, reject) => {
+      const serialPort = CircuitPlaygroundBoard.openSerialPort(this.portName_);
+      const playground = CircuitPlaygroundBoard.makePlaygroundTransport(serialPort);
+      const board = new five.Board({io: playground, repl: false, debug: false});
+      board.once('ready', () => {
+        this.fiveBoard_ = board;
+        resolve();
+      });
+      board.once('error', reject);
+    });
+  }
+
+  /**
+   * Initialize a set of johnny-five component controllers.
+   * Exposed as a separate step here for the sake of the setup page; generally
+   * it'd be better to just call connect(), above.
+   * @throws {Error} if called before connecting to firmware
+   */
+  initializeComponents() {
+    if (!this.fiveBoard_) {
+      throw new Error('Cannot initialize components: Not connected to board firmware.');
+    }
+
+    this.prewiredComponents_ = {
+      board: this.fiveBoard_,
+      ...initializeCircuitPlaygroundComponents(this.fiveBoard_),
+      ...J5_CONSTANTS
+    };
+  }
+
+  /**
+   * Forward events from individual components out to listeners attached to
+   * this EventEmitter.
+   * @private
+   * @throws {Error} if called before connecting to firmware
+   */
+  initializeEventForwarding() {
+    if (!this.fiveBoard_) {
+      throw new Error('Cannot initialize event forwarding: Not connected to board firmware.');
+    }
+
+    this.fiveBoard_.on('disconnect', () => this.emit('disconnect'));
+  }
+
+  /**
+   * Disconnect and clean up the board controller and all components.
+   */
+  destroy() {
+    // Investigate: What do we need to tear down here?
```

Comment:
This comment to be handled in my next PR, expanding on and testing the individual component setup/teardown logic.